anndata/_io/merge.py

```diff
@@ -0,0 +1,647 @@
+
+
+def read_only_dict(group: Union[ZarrGroup, H5Group]):
+    """Read the dict and leave the rest of the group alone."""
+
+    def callback(func, elem_name: str, elem, iospec):
+        if iospec.encoding_type == "dict":
+            return {k: v for k, v in elem.items()}
+        else:
+            return elem
+
+    return read_dispatched(group, callback=callback)
+
+
+def get_encoding_type(group: Union[ZarrGroup, H5Group]) -> str:
+    """Get the encoding type of a group."""
+    return group.attrs.get("encoding-type")
+
+
+def get_shape(group: Union[ZarrGroup, H5Group]) -> str:
+    """Get the shape of a group (array, sparse matrix, etc.)"""
+    if group.attrs["encoding-type"] == "array":
+        return group.shape
+    elif group.attrs["encoding-type"] in SPARSE_MATRIX:
+        return group.attrs.get("shape")
+    raise NotImplementedError(f"Cannot get shape of {group.attrs['encoding-type']}")
+
+
+def _gen_elem_to_append(elems, axis=0, reindexers=None, fill_value=None):
+    for elem, ri in zip(elems, reindexers):
+        if ri.no_change:
+            yield elem
+        else:
+            yield ri(elem, axis=1 - axis, fill_value=fill_value)
+
+
+def gen_reindexers_df_inner(dfs: Sequence[pd.DataFrame], axis: Literal[0, 1] = 0):
+    if axis == 0:
+
+        def df_indices(x):
+            return x.columns
+
+    elif axis == 1:
+
+        def df_indices(x):
+            return x.indices
+
+    # TODO Handle empty later
+    common_ind = reduce(lambda x, y: x.intersection(y), (df_indices(df) for df in dfs))
+    reindexers = [Reindexer(df_indices(df), common_ind) for df in dfs]
+    return reindexers
+
+
+def gen_reindexers_array_inner(
+    arrays: Sequence[Union[ZarrArray, H5Array, SparseDataset]], axis: Literal[0, 1] = 0
+):
+    min_ind = min(a.shape[1 - axis] for a in arrays)
+    reindexers = [
+        gen_reindexer(pd.RangeIndex(min_ind), pd.RangeIndex(a.shape[1 - axis]))
+        for a in arrays
+    ]
+
+    return reindexers
+
+
+def _gen_slice_to_append(
+    datasets: Sequence[SparseDataset],
+    reindexers=None,
+    axis=0,
+    fill_value=None,
+):
+    for ds, ri in zip(datasets, reindexers):
+        n_slices = ds.shape[axis] * ds.shape[1 - axis] // MAX_LOAD_SIZE
+        if n_slices < 2:
+            yield ri(to_memory(ds), axis=1 - axis, fill_value=fill_value)
+        else:
+            slice_size = MAX_LOAD_SIZE // ds.shape[1 - axis]
+            if slice_size == 0:
+                slice_size = 1
+            rem_slices = ds.shape[axis]
+            idx = 0
+            while rem_slices > 0:
+                ds_part = None
+                if axis == 0:
+                    ds_part = ds[idx : idx + slice_size, :]
+                elif axis == 1:
+                    ds_part = ds[:, idx : idx + slice_size]
+
+                yield ri(ds_part, axis=1 - axis, fill_value=fill_value)
+                rem_slices -= slice_size
+                idx += slice_size
+
+
+def _write_concat_sparse(
```

Comment:
It's not clear to me that special casing is happening for concatenation of sparse arrays where no reindexing needs to happen. This would be important for performance.

Reply:
So I did this when there was no reindexing required at all:

```python
class IdentityReindexer:
    def __init__(self):
        self.no_change = True

    def __call__(self, x, *args, **kwargs):
        return x
```

But I wasn't sure if I wanted to do it for the cases where some matrices don't require reindexing. If it is ok to do that, then I think this would be better done by changing the Reindexer class.